pkg/stub/handler_test.go

```diff
@@ -451,6 +453,89 @@ func TestImageStreamEvent(t *testing.T) {
 	validate(true, err, "", cfg, conditions, statuses, t)
 }
 
+func TestImageStreamErrorRetry(t *testing.T) {
+	h, cfg, event := setup()
+	processCred(&h, cfg, t)
+	mimic(&h, x86OCPContentRootDir)
+	err := h.Handle(event)
+	statuses := []corev1.ConditionStatus{corev1.ConditionFalse, corev1.ConditionTrue, corev1.ConditionTrue, corev1.ConditionTrue, corev1.ConditionFalse, corev1.ConditionFalse, corev1.ConditionFalse}
+	validate(true, err, "", cfg, conditions, statuses, t)
+	// expedite the stream events coming in
+	cache.ClearUpsertsCache()
+
+	tagVersion := int64(1)
+	is := &imagev1.ImageStream{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "foo",
+			Annotations: map[string]string{
+				v1.SamplesVersionAnnotation: h.version,
+			},
+		},
+		Spec: imagev1.ImageStreamSpec{
+			Tags: []imagev1.TagReference{
+				{
+					Name:       "foo",
+					Generation: &tagVersion,
+					From: &corev1.ObjectReference{
+						Kind: "DockerImage",
+						Name: "foofoo",
+					},
+				},
+			},
+		},
+		Status: imagev1.ImageStreamStatus{
+			Tags: []imagev1.NamedTagEventList{
+				{
+					Tag: "foo",
+					Conditions: []imagev1.TagEventCondition{
+						{
+							Generation: tagVersion,
+							Status: corev1.ConditionFalse,
+							Message: "failed import",
+						},
+					},
+				},
+			},
+		},
+	}
+
+	h.processImageStreamWatchEvent(is, false)
+
+	if !cfg.ConditionTrue(v1.ImportImageErrorsExist) {
+		t.Fatalf("Import Error Condition not true: %#v", cfg)
+	}
+
+	importError := cfg.Condition(v1.ImportImageErrorsExist)
+	importErrorLastUpdateTime := cfg.Condition(v1.ImportImageErrorsExist).LastUpdateTime
+	h.processImageStreamWatchEvent(is, false)
+	if !importError.LastUpdateTime.Equal(&importErrorLastUpdateTime) {
```

Comment:
i'm a little unclear on the ptr vs copy behavior of a few of these calls...

is cfg.Condition() giving you back a copy, or a reference?  The code above seems to assume it's a reference, but the code below (line 515/516) seems to treat it as a copy.

Reply:
also seems slightly odd to pass a pointer/reference to the equal function when you are trying to compare the values, but i'm assuming that's just how the type's equal signature is written for performance reasons..  true?

Reply:
On the `LastUpdateTime.Equal` calls, yes, that type's equal signature (from k8s/apimachinery) takes a pointer ref for the param.

on cfg.Condition(), yes, it returns a ptr/reference.  Most likely when we first introduced cfg.ConditionUpdate we were returning a copy of the struct.

I'll look into going one way or the other as a separate commit in this PR.  Do you have a preference @bparees as to which direction I go?

Reply:
Sorry I was confused earlier.  The condition array in the Config object is *NOT* an array of pointers to conditions.  It is an array of Conditions.  

The `cfg.Condition` method returns a pointer as an optimization.  But you still need `cfg.ConditionUpdate` to change the original array properly.

You OK with things as they stand @bparees ?  Or would you prefer say `cfg.Condition()` returning a copy vs. a pointer?

Reply:
i think this logic is still wrong:

```
	if !importError.LastUpdateTime.Equal(&importErrorLastUpdateTime) {
```

importError is a pointer to a copy of the condition.

When the actual condition gets updated by calling `processImageStreamWatchEvent`, The copy of the condition pointed to by importError is not going to be changed.  So this check will always succeed because importError is never going to be changed by anything.


The reason i say it's a copy is you got importError from this logic:
```
		for _, rc := range s.Status.Conditions {
			if rc.Type == c {
				return &rc
			}
		}
```

So "rc" will get a copy of the each condition as the for loop iterates, and then you return a pointer to rc.

This is why, as you say, you still have to call cfg.ConditionUpdate to update the original array.